Support for project level metrics and more metric types

roryclaasen opened the pull request:
- Project level metrics can be supplied into the sensor, when its the computers turn they are already added and we skip processing that metric.
- Supported data types are shown in the readme
- Updated original unit tests to support file/project metrics
- Added unit test for project metrics
- Added unit tests for data types

Commit:
Added support for multiple data types and project level metrics

Commit:
Updated readme to include project level measures description

Commit:
Format my changes to match project

Commit:
100% coverage on unit tests

roryclaasen commented on src/test/java/org/sonar/generic/metrics/TestGenericMetricsSensor.java:
```diff
@@ -188,6 +193,6 @@ public void whenExecutingAndCanReadFileAndFileInJsonNotFoundShouldNotPublishMetr
 
     sensor.execute(this.sensorContext);
-    verify(sensorContext, times(0)).newMeasure();
+    verify(sensorContext, times(1)).newMeasure();
```
Even though the file cannot be found, a project level metric is supplied, hence 1 not 0

roryclaasen commented on src/test/java/org/sonar/generic/metrics/TestGenericMetricsSensor.java:
```diff
@@ -129,7 +134,7 @@ public void whenExecutingAndCanReadFileShouldPublishMetrics() throws IOException
     sensor.execute(this.sensorContext);
 
-    verify(sensorContext, times(4)).newMeasure();
+    verify(sensorContext, times(7)).newMeasure();
```
6 file level metrics plus one project level metric

roryclaasen commented on src/main/java/org/sonar/generic/metrics/GenericMetricMeasureComputer.java:
```diff
@@ -26,40 +28,76 @@ public MeasureComputerDefinition define(MeasureComputerDefinitionContext context
 
   @Override
   public void compute(MeasureComputerContext context) {
+    if (context.getMeasure(metric.key()) != null)
```
Generic Metric Sensor adds project measures to the context, this check is to make sure that the computer will not calculate project measures (making the sensor do it)

ericlemes replied:
This doesn't make sense. If your intention is to ignore project metrics, you should add a break in case  PROJECT. This will break the case where you want to compute a sum/average instead of using the provided metrics in the json file. 

I think you are missing a test for a measure computer.
- add a metric which expects the project-level metric to be calculated
- add a metric which expects the project-level metric to be provided in the json file.

Assert if the project metric has the expected value. This will probably highlight a bug here.

ericlemes commented on pom.xml:
```diff
@@ -131,7 +131,7 @@
         </executions>
       </plugin>
-      <plugin>
+       <plugin>
```
Are changes to pom.xml really necessary? They seem to not be changing any behaviour at all.

ericlemes commented on src/main/java/org/sonar/generic/metrics/Environment.java:
```diff
@@ -3,6 +3,7 @@
  * Copyright (C) 2018
  * http://github.com/ericlemes/sonar-generic-metrics
  */
+
```
Is this needed?

ericlemes commented on src/main/java/org/sonar/generic/metrics/EnvironmentImpl.java:
```diff
@@ -3,6 +3,7 @@
  * Copyright (C) 2018
  * http://github.com/ericlemes/sonar-generic-metrics
  */
+
```
Is this needed?

ericlemes commented on src/main/java/org/sonar/generic/metrics/FileReader.java:
```diff
@@ -3,6 +3,7 @@
  * Copyright (C) 2018
  * http://github.com/ericlemes/sonar-generic-metrics
  */
+
```
Is this needed?

ericlemes commented on src/main/java/org/sonar/generic/metrics/GenericMetricMeasureComputer.java:
```diff
@@ -26,40 +28,76 @@ public MeasureComputerDefinition define(MeasureComputerDefinitionContext context
+      computeRatingChildMeasure(context);
+      break;
+    default:
+      System.out.println("Im here");
```
I think you probably forgot this line here.
Also, I think I'd rather "do nothing" than actually throw, since in the first case you still can have "some" metrics, and the second will just interrupt the whole analysis, and in this case, that is the measure computer, on the server side.

roryclaasen replied:
Updated: see comment https://github.com/ericlemes/sonar-generic-metrics/pull/7#issuecomment-477257603

ericlemes commented on src/main/java/org/sonar/generic/metrics/GenericMetrics.java:
```diff
@@ -3,6 +3,7 @@
  * Copyright (C) 2018
  * http://github.com/ericlemes/sonar-generic-metrics
  */
+
```
This file also seems to have no changes at all, just cosmetic ones.

ericlemes commented on src/main/java/org/sonar/generic/metrics/GenericMetricsPlugin.java:
```diff
@@ -11,24 +11,24 @@
 import org.sonar.api.utils.log.Logger;
 
-public class GenericMetricsPlugin implements Plugin{
+public class GenericMetricsPlugin implements Plugin {
```
Same thing for this file. Only cosmetic changes.

ericlemes commented on src/main/java/org/sonar/generic/metrics/GenericMetricsSensor.java:
```diff
@@ -27,74 +30,95 @@
-      context.newMeasure().forMetric(m).on(file).withValue((double)value).save();
-    }
-    else {
+    if (!saveMeasure(context, m, file, value))
```
Good job extracting this in another function!

ericlemes commented on src/test/java/org/sonar/generic/metrics/TestEnvironmentImpl.java:
```diff
@@ -3,6 +3,7 @@
  * Copyright (C) 2018
  * http://github.com/ericlemes/sonar-generic-metrics
  */
+
```
Only cosmetic changes. Please, revert.

ericlemes commented on src/test/java/org/sonar/generic/metrics/TestFileReaderImpl.java:
```diff
@@ -3,6 +3,7 @@
  * Copyright (C) 2018
  * http://github.com/ericlemes/sonar-generic-metrics
  */
+
```
Only cosmetic changes. Please, revert.

ericlemes reviewed (changes requested):
I think the only major concern is the "NotImplementedException". Everything else is cosmetic. 
Overall, I don't mind the changes adding spaces at the end of lines, but as a general rule, if you don't have any functional change on the file, I suggest just reverting it, since will only add unecessary things in the history, making merges more difficult.

Overall, great job!

Commit:
Revert Changes to 245916d16622384b60b07be387f7fcaabaf27d9c

Commit:
Updated Computer & Tests as requested

roryclaasen commented:
Apologies for the cosmetic changes, thats eclipse auto formatting.
Reverted those file changes to 245916d16622384b60b07be387f7fcaabaf27d9c and removed throw in `GenericMetricMeasureComputer.java`

ericlemes reviewed (changes requested):
I think there's a problem with the measure computer. It is not choosing correctly "when" to calculate or not the values. Bear in mind that if you try to add a measure in the project that you already added, you will have an exception.

roryclaasen commented:
> I think there's a problem with the measure computer. It is not choosing correctly "when" to calculate or not the values. Bear in mind that if you try to add a measure in the project that you already added, you will have an exception.

Sonar actually will calculate the other metrics through the `PROJECT` cause. I've just added a break for `PROJECT` and none of the metrics (without project level measures) are shown (image bellow). So that case is still needed.

![image](https://user-images.githubusercontent.com/2751799/55159615-63370480-5159-11e9-96ed-c29fa69867fa.png)

> This doesn't make sense. If your intention is to ignore project metrics, you should add a break in case PROJECT. This will break the case where you want to compute a sum/average instead of using the provided metrics in the json file.

Currently project level metrics are added to Sonar through the **[Sensor](https://github.com/roryclaasen/sonar-generic-metrics/blob/master/src/main/java/org/sonar/generic/metrics/GenericMetricsSensor.java#L73-L83)**. Therefore when the computer runs, the measure already exists on context and hence the if statement. There is no point trying to calculate a measure if one already exists. Without the statement you would receive that exception when it tried to add (another) project measure.

There is no point storing any more data in the metric json as by the time it gets read in (through sensor) and passed to the computer it is striped of everything but the values that sonar uses. I couldn't add a flag to say don't calculate this measure because it jsut wouldn't exist on compute. By adding the measure in sensor then checking if it existed already in computer was the workaround for this problem.

### Additional Tests Requested

#### Add a metric which expects the project-level metric to be calculated

Currently this is being covered by the following tests as none of them have project measures and so they calculate them (and checks that they add them with the correct value):

- `testWhenCallingComputeAndMetricIsFloatAndNoMeasuresShouldComputeSumCorrectly()`
- `testWhenCallingComputeAndComponentIsProjectAndMetricIsFloatShouldComputeSumCorrectly()`
- `testWhenCallingComputeAndComponentIsModuleAndMetricIsFloatShouldComputeSumCorrectly()`
- `testWhenCallingComputeAndComponentIsDirectoryAndMetricIsFloatShouldComputeSumCorrectly()`
- `testWhenCallingComputeAndMetricIsIntAndNoMeasuresShouldComputeSumCorrectly()`
- `testWhenCallingComputeAndComponentIsProjectAndMetricIsIntShouldComputeSumCorrectly()`
- `testWhenCallingComputeAndComponentIsModuleAndMetricIsIntShouldComputeSumCorrectly()`
- `testWhenCallingComputeAndComponentIsDirectoryAndMetricIsIntShouldComputeSumCorrectly()`
- `testWhenCallingComputeAndMetricIsPercentAndNoMeasuresShouldComputeAverageCorrectly()`
- `testWhenCallingComputeAndComponentIsProjectAndMetricIsPercentShouldComputeAverageCorrectly()`
- `testWhenCallingComputeAndComponentIsModuleAndMetricIsPercentShouldComputeAverageCorrectly()`
- `testWhenCallingComputeAndComponentIsDirectoryAndMetricIsPercentShouldComputeAverageCorrectly()`
- `testWhenCallingComputeAndMetricIsRaitingAndNoMeasuresShouldComputeAverageCorrectly()`
- `testWhenCallingComputeAndComponentIsProjectAndMetricIsRatingShouldComputeAverageCorrectly()`
- `testWhenCallingComputeAndComponentIsModuleAndMetricIsRatingShouldComputAverageCorrectly()`
- `testWhenCallingComputeAndComponentIsDirectoryAndMetricIsRatingShouldComputeAverageCorrectly()`

#### Add a metric which expects the project-level metric to be provided in the json file.

Test `testWhenCallingComputeWithMeasureAllReadyAddedShouldNotAddMeasure()` checks to make sure that if another project measure exists then the measure will not be calculated by the computer.
I've added a test in 02fe6cc8bc4305c580222486cb414c5f5552c84c ([TestGenericMetricMeasureComputer.java#L357-L367](https://github.com/roryclaasen/sonar-generic-metrics/blob/master/src/test/java/org/sonar/generic/metrics/TestGenericMetricMeasureComputer.java#L357-L367)) to make sure that it will request the correct value and not the computer calculated one
